File: lang/c/src/vpkppc/format/APK.cpp
#include <vpkppc/format/APK.h>

#include <sourceppc/Helpers.h>

using namespace sourceppc;
using namespace vpkpp;

const uint32_t VPKPP_APK_SIGNATURE = APK_SIGNATURE;
const char* VPKPP_APK_EXTENSION    = APK_EXTENSION.data();

SOURCEPP_API vpkpp_pack_file_handle_t vpkpp_apk_create(const char* path) {
	SOURCEPP_EARLY_RETURN_VAL(path, nullptr);

	auto packFile = APK::create(path);
	if (!packFile) {
		return nullptr;
	}
	return packFile.release();
}

SOURCEPP_API vpkpp_pack_file_handle_t vpkpp_apk_open(const char* path, vpkpp_entry_callback_t callback) {
	SOURCEPP_EARLY_RETURN_VAL(path, nullptr);

	auto packFile = APK::open(path, callback ? [callback](const std::string& entryPath, const Entry& entry) {
		callback(entryPath.c_str(), const_cast<Entry*>(&entry));
	} : static_cast<PackFile::EntryCallback>(nullptr));
	if (!packFile) {
		return nullptr;
	}
	return packFile.release();
}

File: lang/csharp/src/vpkpp/Format/APK.cs
using System;

namespace sourcepp.vpkpp.Format;

using OpenPropertyRequest = Func<PackFile, OpenProperty, byte[]>;

using EntryCallback = Action<string, Entry>;

public class APK : PackFile
{
	protected APK(nint handle, bool managed = true) : base(handle, managed)
	{
	}

	public static APK? Create(string path)
	{
		var handle = DLL.vpkpp_apk_create(path);
		return handle == nint.Zero ? null : new APK(handle);
	}

	public new static APK? Open(string path, EntryCallback? callback = null, OpenPropertyRequest? _ = null)
	{
		var handle = DLL.vpkpp_apk_open(path, callback is not null ? (entryPath, entry) =>
		{
			callback(entryPath, new Entry(entry, false));
		} : null);
		return handle == nint.Zero ? null : new APK(handle);
	}
}
